### ISSUE TITLE
TC39 Temporal support

### ISSUE THREAD

**Issue by LinusU** [prisma/prisma#16119]
## Problem

Working with temporal data in JavaScript is currently quite hard. The builtin `Date` class only covers limited use cases, and have bugs which cannot be fixed whilst still conforming to specification. Using the `Date` class to represent a specific date (instead of a date and time) will lead to bug such as [peoples birthdays being off-by-one](https://github.com/prisma/prisma/issues/4355#issuecomment-1010513913). Using the `Date` class to represent a specific time (without a date) can lead to the hour being changed to basically any value, depending on the timezone of the database, the server, and the client.

We would like to remove `Date` completely from our codebase.

## Suggested solution

The TC39 has developed [a new standard that fixes these issues](https://github.com/tc39/proposal-temporal). There is implementations available in [v8 (behind `--harmony-temporal`)](https://bugs.chromium.org/p/v8/issues/detail?id=11544) and in Safari (behind `--useTemporal`). There is also a [polyfill made by the people behind the proposal](https://github.com/js-temporal/temporal-polyfill).

In this standard there are several classes that can be used to represent temporal data. I would like to see support for the following classes in Prisma:

Temporal | MySQL | PostgresQL
-------- | ----- | ----------
`Instant` | `TIMESTAMP` | <sup>**1**</sup>
`ZonedDateTime` | <sup>**2**</sup> | `timestamp with time zone`
`PlainDate` | `DATE` | `date`
`PlainTime` | `TIME` | `time`
`PlainDateTime` | `DATETIME` | `timestamp`

<sup>**1**</sup> This could be emulated using `bigint` and represeting nanoseconds since unix epoch.
<sup>**2**</sup> This could be emulated using two fields, one `DATETIME` and one `VARCHAR` for the timezone.

I think that we should use the same names in Prisma as in the Temporal standard.

I also think that we should deprecate the current `DateTime`, which is saved without timezone information in the database, but is the deserialized into the `Date` class which has a timezone.

## Alternatives

An alternative could be to have some opt-in flag that switches out the `Date` class with the `PlainDateTime` class from Temporal.

## Additional context

We are currently migrating our codebase to use Prisma, and the state of the date support is making me a bit worried that we will introduce subtle bugs. Mu guess is that our code currently only works because both our database and our server is configured to use the UTC timezone.

After doing the research on all of this I think that we'll switch to using a `bigint` field to represent nanoseconds, and then convert that to `Temporal.Instant` in our codebase, since we don't currently have a use case for storing timestamps in different timezones. But this will be an extra headache for us, and I think that it would be great if Prisma could handle this for us.

**Comment by SevInf** [comment]
This is something we would definitely like to explore in the future, however, at this moment proposal is not finalized — current text explicitly discourages implementors from shipping it unflagged and thuse, it is not available anywhere without a feature flag. 
Shipping a polyfill is also pretty risky at the moment — changes to the proposal, even though not very likely, are still possible. So we'd might end up in a situation where we would ship an outdated polyfill, that does not match  native behaviour in the end.
I the safest way to approach is to wait until proposal reaches at least Stage 4 and have unflagged implementations available.

**Comment by LinusU** [comment]
@SevInf yeah, I should probably have elaborated on that! I'm not sure that I think that Prisma should ship a polyfill, in fact I'm quite certain that it shouldn't.

-----

A quick update on the Temporal status for people not familiar:

- [x] The IETF needs to approve [an update to RFC3339](https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/). It's currently "In WG Last Call" and is scheduled to be discussed at IETF-115 which will happen 5 - 11 november. This is tracked in this issue:
    https://github.com/tc39/proposal-temporal/issues/1450
- [x] Two browsers will ship unflagged implementations. My personal guess is Chrome and Safari, since they already have flagged implementations. But Firefox also has [a WIP patch-set in this issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1519167)
- [x] The proposal will advance to Stage 4

Another interesting thing of note is that Temporal is already available in Node.js, behind the `--harmony_temporal` flag. But of course, as with every harmony flag, this is unstable.

-----

What I personally would like to see, and would be happy to contribute patches to, would be a [Prisma Client feature preview](https://www.prisma.io/docs/concepts/components/preview-features/client-preview-features) that would enable the above described types. It would be up to the user to provide support for Temporal, be that by using the `--harmony_temporal`, or by using a pollyfill.

Since I'm quite new to Prisma, I'm not certain exactly how the development process looks. But as stated, I'd would be very happy to help with this. I'm a bit hesitant to go thru with our migration before we have at least a stop gap solution that would guarantee that we won't have any data corruption, since we have a lot of temporal fields, used in a number of ways.

**Comment by trevoro** [comment]
An abstraction of this particular feature would be a plugin system or middleware that would let developers add types or other sugar in a way that is non-standard but still allows people to take advantage of the Prisma workflow and ecosystem. For example in this case we could have some middleware and schema awareness that lets us turn a `TIMESTAMP WITHOUT TIMEZONE` into a `Temporal.Instant`. Maybe?

**Comment by psupina** [comment]
> `ZonedDateTime`	**2**	`timestamp with time zone`

Beware that in PostgreSQL, "timestamp with time zone" column does not actually store any time zone information, it stores the date time in the UTC time zone.
Therefore you would need the similar workaround with two columns as in the solution for MySQL.

**Comment by jthoward64** [comment]
As an update, Temporal is now stage 3; might not be a bad idea to at least set up an RFC for a proposed API or something

**Comment by Akxe** [comment]
I think that instead of full integration of `Temporal` into Prisma, how about adding the ability to add custom type transformation to Prisma? This way I would be able to add the polyfill that I use in my codebase, or potentially add even some other types!

PS: This would solve the timezone info because I could encode the timezone logic into logic in the type transform.

**Comment by gforge** [comment]
It seems that Temporal just hit stage 4 - https://github.com/tc39/proposal-temporal

I don't know how many hours I've spent over the years fighting datetime when all I wanted was date, it would be great if this could be reactivated

**Comment by jthoward64** [comment]
> It seems that Temporal just hit stage 4 - https://github.com/tc39/proposal-temporal
> 
> I don't know how many hours I've spent over the years fighting datetime when all I wanted was date, it would be great if this could be reactivated

Worth noting typescript has added native types as well as of v6.

**Comment by rejc2** [comment]
Temporal API is now enabled by default in [Node 26](https://nodejs.org/en/blog/release/v26.0.0#temporal-api).

(I too would like to be able to forget that `Date` ever existed ;)

**Comment by davidjbng** [comment]
While prisma could do more to push this forward, I believe this has to be added to `node-postgres` first. Which I'm following here:
- https://github.com/brianc/node-postgres/issues/3663

**Comment by sina-heidariaan** [comment]
If you need this today: temporal-sql maps Postgres timestamptz/timestamp/date/time/timetz/interval to TC39 Temporal (Instant/PlainDateTime/PlainDate/PlainTime/Duration) for pg / postgres.js / Drizzle / Prisma. It handles the fiddly bits — interval⇄Duration and microsecond precision — and keeps Date out entirely. npm: temporal-sql · repo: github.com/sina-heidariaan/temporal-sql
